packages/next/src/plugins/plugin.ts

```diff
@@ -0,0 +1,229 @@
+import {
+  CreateDependencies,
+  CreateNodes,
+  CreateNodesContext,
+  NxJsonConfiguration,
+  TargetConfiguration,
+  readJsonFile,
+  writeJsonFile,
+} from '@nx/devkit';
+import { dirname, join } from 'path';
+
+import { readTargetDefaultsForTarget } from 'nx/src/project-graph/utils/project-configuration-utils';
+import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
+import { existsSync, readdirSync } from 'fs';
+
+import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';
+import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
+import { type NextConfig } from 'next';
+import { PHASE_PRODUCTION_BUILD } from 'next/constants';
+
+export interface NextPluginOptions {
+  buildTargetName?: string;
+  devTargetName?: string;
+  startTargetName?: string;
+}
+
+const cachePath = join(projectGraphCacheDirectory, 'next.hash');
+const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};
+
+const calculatedTargets: Record<
+  string,
+  Record<string, TargetConfiguration>
+> = {};
+
+function readTargetsCache(): Record<
+  string,
+  Record<string, TargetConfiguration>
+> {
+  return readJsonFile(cachePath);
+}
+
+function writeTargetsToCache(
+  targets: Record<string, Record<string, TargetConfiguration>>
+) {
+  writeJsonFile(cachePath, targets);
+}
+
+export const createDependencies: CreateDependencies = () => {
+  writeTargetsToCache(calculatedTargets);
+  return [];
+};
+
+// TODO(nicholas): Add support for .mjs files
+export const createNodes: CreateNodes<NextPluginOptions> = [
+  '**/next.config.{js, cjs}',
+  async (configFilePath, options, context) => {
+    const projectRoot = dirname(configFilePath);
+
+    // Do not create a project if package.json and project.json isn't there.
+    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
+    if (
+      !siblingFiles.includes('package.json') &&
+      !siblingFiles.includes('project.json')
+    ) {
+      return {};
+    }
+
+    options = normalizeOptions(options);
+
+    const hash = calculateHashForCreateNodes(projectRoot, options, context);
```

Comment:
We also need to take lockfile into account. See https://github.com/nrwl/nx/blob/master/packages/cypress/src/plugins/plugin.ts#L74-L76